kubernetes/deployments_metrics_blackbox_test.go

```diff
@@ -0,0 +1,394 @@
+}
+
+func TestGetNetworkSentRange(t *testing.T) {
+	testCases := metricRangeTestCases
+	test := &testHawkularGetter{}
+	config := &kubernetes.MetricsClientConfig{
+		MetricsURL:     "myMetricsServer",
+		BearerToken:    "token",
+		HawkularGetter: test,
+	}
+	client, err := kubernetes.NewMetricsClient(config)
+	require.NoError(t, err, "Failed to create metrics client")
+
+	for _, testCase := range testCases {
+		test.input = testCase
+		metrics, err := client.GetNetworkSentMetricsRange(testCase.pods, testCase.namespace, testCase.startTime, testCase.endTime, 0)
+		if !assert.NoError(t, err, "Getting network metrics failed") ||
+			!assert.NotNil(t, metrics, "Nil result from network metrics") {
+			continue
+		}
+
+		// Check that the result has the correct value and timestamp and that the Hawkular API was called
+		// with the expected values
+		output := test.result.output
+		verifyMetrics(metrics, testCase, output, "network/tx_rate", t)
+
+		// Verify the remaining filters
+		verifyMetricRangeFilters(testCase, output.filters, t)
+	}
+}
+
+func verifyMetrics(metrics []*app.TimedNumberTuple, testCase *testMetricsInput, result *testMetricsOutput,
+	gaugeDesc string, t *testing.T) {
+	// If limit is specified, check that the number of metrics doesn't exceed that limit
+	numMetrics := len(metrics)
+	if testCase.limit > 0 && !assert.True(t, numMetrics <= testCase.limit, "Too many metrics returned") {
+		return
+	}
+
+	// Check that the result has the correct values and timestamps
+	require.True(t, numMetrics <= len(testCase.buckets), "More metrics than buckets") // Sanity check
+	for i := 0; i < numMetrics; i++ {
+		// Iterate backwards since earlier buckets may have been discarded due to limit parameter
+		metric := metrics[numMetrics-1-i]
+		bucket := testCase.buckets[len(testCase.buckets)-1-i]
+		if !assert.NotNil(t, metric.Value, "Nil value in network metric") ||
+			!assert.InEpsilon(t, bucket.Avg, *metric.Value, fltEpsilon, "Incorrect value in network metric") ||
+			!assert.NotNil(t, metric.Time, "Nil time in network metric") ||
+			!assert.InEpsilon(t, hawkular.ToUnixMilli(bucket.Start), *metric.Time, fltEpsilon, "Incorrect time in network metric") {
+			return
+		}
+	}
+
+	// Check that ReadBuckets was called with the correct inputs
+	if !assert.Equal(t, testCase.namespace, result.namespace, "ReadBuckets called with incorrect namespace") ||
+		!assert.Equal(t, hawkular.Gauge, result.metricType, "Incorrect Hawkular metric type") {
+		return
+	}
+
+	// Check that the tags used in the Hawkular query are correct
+	uids := make([]string, len(testCase.pods))
+	for idx := range testCase.pods {
+		uids[idx] = string(testCase.pods[idx].UID)
+	}
+	expectedPodTag := strings.Join(uids, "|")
+	expectedTags := map[string]string{
+		"descriptor_name": gaugeDesc,
+		"type":            "pod",
+		"pod_id":          expectedPodTag,
+	}
+	tags := tagsToMap(result.filters.Get("tags"), t)
+	if t.Failed() {
+		return
+	}
+	for key, value := range expectedTags {
```

Comment:
I haven't looked at the rest of this code but I tend to do testing a little bit different. To me it seems that the order of expected tags does matter when really it shouldn't, right? To test for something where the order doesn't matter I suggest this approach as a rough sketch:

```go
// tags to be found
toBeFound := map[string]struct{} {
  "foo": {},
  "bar": {}, 
}
for _, tag := range tags {
  _, ok := toBeFound[tag]
  require.True(t, ok, "found unexpected tag: %s", tag)
  delete(toBeFound, tag)
}
require.Empty(t, toBeFound, "failed to find these tags: %+v", toBeFound)
```
I know that a `map` looks weird but it is much easier to use a map than to use an array. @xcoulon knows this test method as well.

Reply:
Good catch, I'll forward this.  Order should never matter.